### PR TITLE
Norm subgradient at 0

### LINKED ISSUE #2421: Gradient of zero norm is nan
If a norm is zero, its gradient returns nan:

```python
x = Variable(torch.zeros(1), requires_grad=True)
x.norm().backward()
print x.grad

# Variable containing:
# nan
# [torch.FloatTensor of size 1]
```

Obviously just happening because the gradient divides by the norm, but the (sub)gradient here should probably be zero, or at least not nan, since that will propagate to make all updates nan.  Probably low priority, as it's not going to be an issue in 99% of cases, but we're doing a few things with (exact) line searches where this caused a nan to appear, breaking everything downstream.

### PR DESCRIPTION
**Pull request by albanD**
Fix #2421 

Note that with this, at `0`, it is now different to use `.norm()` or `.pow(2).sum().sqrt()` as the first one will return `0` and the second one `NaN`

### CODE REVIEW

File: torch/autograd/_functions/reduce.py [hunk 1]
```diff
@@ -226,43 +226,35 @@ def forward(ctx, input, p=2, dim=None, keepdim=None):
...
+            input_pow = input.abs().pow(ctx.p - 2)
+            output_pow = output.pow(ctx.p - 1)
+            grad_input = input.mul(input_pow).mul(grad_output).div(output_pow)
 
-            if ctx.keepdim is False and input.dim() != 1:
-                grad_output = grad_output.unsqueeze(ctx.dim)
-                output = output.unsqueeze(ctx.dim)
+        # Special case at 0 where we return a subgradient containing 0
+        zero_mask = (output == 0).type_as(grad_input)
+        grad_input.mul(1 - zero_mask)
```
Comment (fmassa) [inline torch/autograd/_functions/reduce.py]: Did you intent to do inplace multiplication?

Comment (albanD) [reply torch/autograd/_functions/reduce.py]: Yes, my bad, the test was wrong... Thanks for spotting that.

**Comment by soumith** [comment]
thanks Alban!
